numpyro/infer/util.py

```diff
@@ -416,49 +415,102 @@ def single_chain_init(key):
     return init_params, potential_fn, constrain_fun
 
 
-def predictive(rng_key, model, posterior_samples, *args, num_samples=None, return_sites=None, **kwargs):
+def _predictive(rng_key, model, posterior_samples, num_samples, return_sites=None,
+                parallel=True, model_args=(), model_kwargs={}):
+    rng_keys = random.split(rng_key, num_samples)
+
+    def single_prediction(val):
+        rng_key, samples = val
+        model_trace = trace(seed(substitute(model, samples), rng_key)).get_trace(
+            *model_args, **model_kwargs)
+        if return_sites is not None:
+            if return_sites == '':
+                sites = {k for k, site in model_trace.items() if site['type'] != 'plate'}
+            else:
+                sites = return_sites
+        else:
+            sites = {k for k, site in model_trace.items()
+                     if site['type'] != 'plate' and k not in samples}
+        return {name: site['value'] for name, site in model_trace.items() if name in sites}
+
+    if parallel:
+        return vmap(single_prediction)((rng_keys, posterior_samples))
+    else:
+        return lax.map(single_prediction, (rng_keys, posterior_samples))
+
+
+class Predictive(object):
     """
-    Run model by sampling latent parameters from `posterior_samples`, and return
-    values at sample sites from the forward run. By default, only sample sites not contained in
-    `posterior_samples` are returned. This can be modified by changing the `return_sites`
-    keyword argument.
+    This class is used to construct predictive distribution. The predictive distribution is obtained
+    by running model conditioned on latent samples from `posterior_samples`.
 
     .. warning::
-        The interface for the `predictive` function is experimental, and
+        The interface for the `Predictive` class is experimental, and
         might change in the future.
 
-    :param jax.random.PRNGKey rng_key: seed to draw samples
     :param model: Python callable containing Pyro primitives.
     :param dict posterior_samples: dictionary of samples from the posterior.
-    :param args: model arguments.
+    :param callable guide: optional guide to get posterior samples of sites not present
+        in `posterior_samples`.
+    :param dict params: dictionary of values for param sites of model/guide.
+    :param int num_samples: number of samples
     :param list return_sites: sites to return; by default only sample sites not present
         in `posterior_samples` are returned.
-    :param int num_samples: number of samples
-    :param kwargs: model kwargs.
+    :param bool parallel: whether to predict in parallel using JAX vectorized map :func:`jax.vmap`.
```

Comment:
Do you think we should make this `True` by default? How often have you run into memory issues with `vmap` used for doing predictions?

Reply:
This is already True by default. :D

Reply:
Sorry, I meant `False` (I think the corresponding argument in Pyro is `sequential` which is the opposite of what we have here). Sequential will be slow but will almost always return correct results instead of running out of memory.

Reply:
I only face this problem in covtype example. It would be fine to change the default value to False because either way is fast. :D

Reply:
If its fast, let us change it to `False` then to preserve the same behavior as in Pyro.

Reply:
Yeah, I will do it before the release. :)